### PR TITLE
Final stage config isolation support write only

### PR DESCRIPTION
**Pull request by ulysses-you**
### _Why are the changes needed?_

Detect and inject a tag if plan is for writing, then skip doing final stage isolation at query preparation phase.

To make final stage config more flexible with complex Spark application.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
It surprises me that it did not only support for the write side

**Comment by ulysses-you** [comment]
@yaooqinn It is a long time issue but I have not been talked about it something wrong..

File: extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala [hunk 1]
```diff
@@ -58,6 +58,16 @@ case class FinalStageConfigIsolation(session: SparkSession) extends Rule[SparkPl
...
 
     if (isFinalStage(plan)) {
+      // We can not get the whole plan at query preparation phase to detect if current plan is
+      // for writing, so we depend on a tag which is been injected at post resolution phase.
+      // Note: we should still do clean up previous config for non-final stage to avoid such case:
+      // the first statement is write, but the second statement is query.
+      if (conf.getConf(FINAL_STAGE_CONFIG_ISOLATION_WRITE_ONLY) &&
+        conf.getConf(KyuubiSQLConf.MARK_NUM_OUTPUT_COLUMNS) &&
+        !MarkNumOutputColumnsRule.isWrite(session, plan)) {
+        return plan
```
Comment (ulysses-you) [inline extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala:68]: It's real hard to detect if current plan is inside a subquery or a cached plan, so I just check output columns for writing. Most of case should be covered.

**Comment by ulysses-you** [comment]
cc @yaooqinn @cfmcgrady

File: extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala [hunk 2]
```diff
@@ -58,6 +58,16 @@ case class FinalStageConfigIsolation(session: SparkSession) extends Rule[SparkPl
     }
 
     if (isFinalStage(plan)) {
+      // We can not get the whole plan at query preparation phase to detect if current plan is
+      // for writing, so we depend on a tag which is been injected at post resolution phase.
+      // Note: we should still do clean up previous config for non-final stage to avoid such case:
+      // the first statement is write, but the second statement is query.
+      if (conf.getConf(FINAL_STAGE_CONFIG_ISOLATION_WRITE_ONLY) &&
+        conf.getConf(KyuubiSQLConf.MARK_NUM_OUTPUT_COLUMNS) &&
```
Comment (cfmcgrady) [inline extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala:66]: Rule `MarkNumOutputColumnsRule` doesn't make any sense when condition `FINAL_STAGE_CONFIG_ISOLATION_WRITE_ONLY` is false, maybe we can simplify these two config?

Comment (ulysses-you) [reply extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala]: make sense, removed config `MARK_NUM_OUTPUT_COLUMNS`

**Comment by ulysses-you** [comment]
@yaooqinn @cfmcgrady any comments ?

**Review by yaooqinn — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks for review, merging to master

**Pull request closed without merging**